File: samples/semgrep/bad.py
# Layer 2 bait for Python — issue #21.
#
# Python shipped as a full Layer 1 language (Ruff's 13 families plus mypy
# strict) and Layer 2 never followed: `semgrep --config semgrep samples/python`
# reported `Ran 19 rules on 0 files`. Not one rule id could match a `.py` file.
#
# What is here is only the measured gap. Ruff already covers five of the twelve
# conventions outright — T201, SIM105, S324, S608, S602 — and ASYNC251 covers
# sync-over-async, so none of those get a Semgrep rule they do not need. See
# docs/EVAL-vs-oss-tools.md §1f.
#
# Every secret-shaped string in this file is planted bait and has never been
# valid anywhere. See the README next to it.
#
# Nothing here is compiled, linted or type-checked — `ruff check` and `mypy`
# run against samples/python/, not this directory, so Layer 2 bait can never
# shift a Layer 1 count. Semgrep only parses it.
import datetime
import logging
import time
from datetime import date

logger = logging.getLogger(__name__)


# --- todo-without-issue -----------------------------------------------------
# The rule is `languages: [generic]` and always could have read Python; only its
# `paths.include` list kept it out. `#` had to join the comment markers.
# TODO: rename this before the migration


# NEGATIVE CONTROL. A tracked TODO is a decision somebody can find again. `#` is
# both the comment marker and the issue sigil, so this is the case that proves
# the exemption is anchored on `#[0-9]+` and not merely on `#`.
# TODO(#412): tracked, and therefore fine


# --- hardcoded-secret-python ------------------------------------------------
# Branch: plain assignment. Ruff's S105 also catches this one; it is here so the
# rule's own coverage is provable rather than inferred from Ruff's.
API_TOKEN = "sk-live-4f9a2c7e1b8d3a6f"

# Branch: annotated assignment. THE CASE THIS RULE EXISTS FOR — ruff 0.16.1 is
# silent here (measured 2026-08-02), because S105's name list does not reach
# `connection_string`, so a userinfo URL in a Python service was invisible to
# every layer this baseline runs.
CONNECTION_STRING: str = "postgres://admin:hunter2@db.internal:5432/app"


class Client:
    def __init__(self) -> None:
        # Branch: attribute assignment. The value is the base64 body with no
        # spaces in it, and that is not cosmetic — the fourth value guard below
        # drops anything containing whitespace, so a PEM header written out in
        # full would have made this branch look covered while testing the guard
        # instead of the branch.
        self.private_key = "MIIEvQIBADANBgkqhkiG9w0BAQEFAASCBKcwggSjAgEAAoIBAQ"

        # NEGATIVE CONTROL for the value guard, and the reason the guard exists
        # at all: Phase A scored the name test 0 true positives out of 5 (#17).
        # A bare URL is an endpoint, not a credential — and this one is in a
        # `token`-named attribute, which is exactly the false positive observed.
        self.token_endpoint = "https://auth.example.com/oauth2/token"

        # NEGATIVE CONTROL. Under 12 characters is a sentinel, not a key.
        self.api_key = "none"

        # NEGATIVE CONTROL. An obvious placeholder is not a secret.
        self.client_secret = "changeme"


# NEGATIVE CONTROL for the fourth value guard, the one the TS and C# twins do
# not have. Measured against 4,133 files of Django, Celery, SQLAlchemy, Flask
# and httpx, prose in a secret-named constant was the largest single false-
# positive class the other three guards left. A credential does not contain a
# space.
SECRET_KEY_WARNING = "Your SECRET_KEY is missing and the server will not start."


# --- no-ambient-clock-python ------------------------------------------------
# Branch: `.now()`. The fully-qualified spelling `import datetime` gives you,
# which is the one most Python code uses. It is baited because the first draft
# of the rule assumed semgrep matches a dotted suffix and would cover this from
# `datetime.now(...)`. It does not, and this fixture is what said so.
def expires_at() -> datetime.datetime:
    return datetime.datetime.now() + datetime.timedelta(days=1)


# Branch: `.utcnow()`.
def created_at() -> datetime.datetime:
    return datetime.datetime.utcnow()


# Branch: `.today()` — and the OTHER half of the receiver regex. The two cases
# above go through its `datetime.` prefix; this one goes through the prefix
# being absent, which is what `from datetime import date` produces. Both
# alternatives are reached, so neither can be deleted unobserved.
def billing_day() -> date:
    return date.today()


# Branch: `time.time()`, which has no receiver to match on and is its own
# pattern.
def stamp() -> float:
    return time.time()


# NEGATIVE CONTROL. Stamping a timestamp into a log line is not untestable
# business logic — the same carve-out the TS and C# rule makes.
def log_progress(step: str) -> None:
    logger.info("%s at %s", step, datetime.datetime.now())


# --- no-float-for-money-python ----------------------------------------------
# Branch: annotated assignment.
total_amount: float = 0.0


# Branch: bare annotation on a class attribute.
class Invoice:
    balance_due: float


# Branch: function parameter.
def apply_discount(item: str, unit_price: float) -> str:
    return item


# Branch: async function parameter.
async def charge(customer: str, total_fee: float) -> None:
    return None


# NEGATIVE CONTROLS. A float that is not money is not this rule's business, and
# a money amount that is already a Decimal is the fix, not the bug.
def resize(image: str, scale_factor: float) -> str:
    return image


def refund(amount: "decimal.Decimal") -> None:
    return None


# KNOWN GAP, kept visible on purpose. Python only has a type where somebody
# wrote an annotation, so an unannotated local is not reachable by this rule and
# this line is SILENT. A rule keyed on the literal instead would fire on every
# float in the codebase. If a future change makes this reachable, the finding
# has to be added to samples/expected/semgrep.json in the same change.
def running_total(rows: list[float]) -> float:
    total_price = 0.0
    for row in rows:
        total_price += row
    return total_price


# --- no-permission-denied-for-invisible-resource-python ---------------------
class HTTPException(Exception):
    def __init__(self, status_code: int, detail: str = "") -> None:
        super().__init__(detail)


class PermissionDenied(Exception):
    pass


class JSONResponse:
    def __init__(self, status_code: int, content: object = None) -> None:
        self.status_code = status_code


def HttpResponseForbidden(*args: object) -> object:  # noqa: N802
    return object()


# Branch: `is None` + raise with status_code=403 (FastAPI / Starlette).
def get_document(doc_id: str, repo: object) -> object:
    doc = repo.find(doc_id)
    if doc is None:
        raise HTTPException(status_code=403, detail="forbidden")
    return doc


# Branch: falsiness check + raise with status_code=403.
def get_invoice(invoice_id: str, repo: object) -> object:
    invoice = repo.find(invoice_id)
    if not invoice:
        raise HTTPException(status_code=403, detail="forbidden")
    return invoice


# Branch: `is None` + returned response object.
def get_report(report_id: str, repo: object) -> object:
    report = repo.find(report_id)
    if report is None:
        return JSONResponse(status_code=403, content={"detail": "forbidden"})
    return report


# Branch: falsiness check + Django's HttpResponseForbidden, which carries no
# number at all.
def get_page(page_id: str, repo: object) -> object:
    page = repo.find(page_id)
    if not page:
        return HttpResponseForbidden("forbidden")
    return page


# Branch: `is None` + DRF's PermissionDenied — the closest twin of the C# rule's
# `StatusCode.PermissionDenied`.
def get_folder(folder_id: str, repo: object) -> object:
    folder = repo.find(folder_id)
    if folder is None:
        raise PermissionDenied("forbidden")
    return folder


# NEGATIVE CONTROL. 404 for a resource the caller cannot see is the fix, and it
# must not fire.
def get_secret_note(note_id: str, repo: object) -> object:
    note = repo.find(note_id)
    if note is None:
        raise HTTPException(status_code=404, detail="not found")
    return note

File: samples/semgrep/Bad.java
/*
 * INTENTIONALLY BAD CODE — DO NOT FIX.
 *
 * Layer 2 sample: bait for the semgrep/ ruleset. Deliberately kept OUT of the
 * samples/java project so that adding Semgrep bait here never changes the
 * analyzer sample's expected finding count.
 *
 * This file is never compiled — Semgrep only parses it. It does not even sit in
 * a Maven source root, which is why it can carry imports that resolve to
 * nothing.
 *
 * WHAT IS NOT HERE, AND WHY. Three conventions are covered by Error Prone
 * instead, measured 2026-08-09 rather than assumed, so a Semgrep rule for them
 * would be a second finding on one line rather than more coverage:
 *
 *   catch-and-swallow          -> [EmptyCatch], on by default
 *   new Date()                 -> [JavaUtilDate], on by default
 *   LocalDate/LocalDateTime.now() -> [JavaTimeDefaultTimeZone], on by default
 *
 * The ambient-clock rule below is scoped to exactly the two shapes Error Prone
 * does NOT reach — Instant.now() is zone-independent so JavaTimeDefaultTimeZone
 * allows it, and System.currentTimeMillis() has no check at all.
 */
package dev.maximalcode.semgrepsample;

import java.math.BigDecimal;
import java.security.MessageDigest;
import java.sql.Connection;
import java.time.Instant;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import javax.crypto.Cipher;
import org.springframework.jdbc.core.simple.JdbcClient;
import reactor.core.publisher.Mono;

public final class Bad {

    private Bad() {}

    // --- no-float-for-money-java -------------------------------------------
    public static class Invoice {
        double totalAmount;
        float discountAmount;

        // NOT flagged: BigDecimal is the right type.
        BigDecimal netTotal;
    }

    // --- hardcoded-secret-java ---------------------------------------------
    private static final String API_TOKEN = "ghp_A1b2C3d4E5f6G7h8I9j0";
    // A URL is exempt, but NOT when it carries userinfo — same pair as Bad.cs.
    private static final String CONNECTION_STRING = "postgres://admin:hunter2is@db.internal:5432/prod";

    // Negative controls for #17 — these must stay SILENT.
    private static final String TOKEN_ENDPOINT = "https://oauth2.googleapis.com/token";
    private static final String TOKEN_NONE = "none";

    // --- todo-without-issue -------------------------------------------------
    // TODO: switch this over to the new pricing service
    // TODO(#918): this one is fine — it has an issue and must NOT be flagged

    // --- no-ambient-clock-java ----------------------------------------------
    public static boolean isExpired(Instant expiresAt) {
        return expiresAt.isBefore(Instant.now());
    }

    public static long stamp() {
        return System.currentTimeMillis();
    }

    // --- no-float-for-money-java (parameter) --------------------------------
    public static double applyTax(double amount) {
        return amount * 1.19;
    }

    // --- weak-crypto-java ---------------------------------------------------
    public static MessageDigest fingerprint() throws Exception {
        return MessageDigest.getInstance("MD5");
    }

    // Case-insensitivity and the real DES spellings, the same evasions the
    // TypeScript rule was caught missing.
    public static MessageDigest fingerprintUpper() throws Exception {
        return MessageDigest.getInstance("SHA-1");
    }

    public static Cipher legacyCipher() throws Exception {
        return Cipher.getInstance("DESede/CBC/PKCS5Padding");
    }

    // NEGATIVE CONTROL for weak-crypto-java — must stay silent.
    public static MessageDigest strongFingerprint() throws Exception {
        return MessageDigest.getInstance("SHA-256");
    }

    // --- sql-string-concat-java ---------------------------------------------
    public static void findUser(JdbcClient jdbc, String id) {
        jdbc.sql("SELECT * FROM users WHERE id = " + id).query();
    }

    public static void findUserTemplate(org.springframework.jdbc.core.JdbcTemplate t, String id) {
        t.query("SELECT * FROM users WHERE id = " + id, (rs, n) -> null);
    }

    public static void deleteUser(org.springframework.jdbc.core.JdbcTemplate t, String id) {
        t.update("DELETE FROM users WHERE id = " + id);
    }

    public static void executeRaw(Connection conn, String id) throws Exception {
        conn.createStatement().executeQuery("SELECT * FROM users WHERE id = " + id);
    }

    // NEGATIVE CONTROL for the SQL group. A bound parameter must NOT fire, or
    // the rule is "any query call" and gets switched off.
    public static void findUserSafe(JdbcClient jdbc, String id) {
        jdbc.sql("SELECT * FROM users WHERE id = :id").param("id", id).query();
    }

    // --- command-injection-java ---------------------------------------------
    public static Process archive(String path) throws Exception {
        return Runtime.getRuntime().exec("tar -czf backup.tgz " + path);
    }

    public static Process archiveBuilder(String path) throws Exception {
        return new ProcessBuilder("sh", "-c", "tar -czf backup.tgz " + path).start();
    }

    // NEGATIVE CONTROL: an argument array is the fix the message asks for.
    public static Process archiveSafe(String path) throws Exception {
        return new ProcessBuilder("tar", "-czf", "backup.tgz", path).start();
    }

    // --- sync-over-async-java -----------------------------------------------
    // The receiver names are load-bearing: join()/get() are guarded by a name
    // heuristic because Optional.get() and Map.get() are everywhere. Both
    // branches of that guard are baited.
    public static String loadBlocking(CompletableFuture<String> resultFuture) {
        return resultFuture.join();
    }

    public static String loadBlockingGet(CompletableFuture<String> resultFuture) throws Exception {
        return resultFuture.get();
    }

    // The Reactor branch, which carries NO name guard — block() exists on Mono
    // and Flux and essentially nowhere else. Named `m` on purpose, so this
    // fixture fails if somebody "tidies up" by folding the two branches back
    // into one name-guarded pattern.
    public static String loadBlockingReactive(Mono<String> m) {
        return m.block();
    }

    // NEGATIVE CONTROL — Optional.get() is not blocking on anything, and is
    // why the name guard exists at all. Must stay SILENT.
    public static String unwrap(Optional<String> maybe) {
        return maybe.get();
    }

    // --- debug-print-left-behind-java ---------------------------------------
    public static void trace(String message) {
        System.out.println(message);
    }

    public static void traceErr(String message) {
        System.err.println(message);
    }

    public static void traceStack(Exception e) {
        e.printStackTrace();
    }
}
